### PR TITLE
feat: add `tracing` infrastructure to contract runtime

### PR DESCRIPTION
**Pull request by matklad**
(builds on top of #3973)

`tracing` is a generic instrumentation framework. It allows manually
instrumented code to signal point-in-time events and nested spans.
Separately, a consumer for those spans and events can be defined.

In this particular case, we use `tracing` to measure time of various
phases of contract execution, with the output ending up like this (read
it bottom up):

        398.15ms compile_module
      398.19ms compile_module_cached_wasmer_impl
      140.19µs run_wasmer/call
    407.21ms run_wasmer

The output's format can certainly be improved, but the beauty of
`tracing` is that it can be done separately from instrumentation.

`tracing` is a bit over-engineered to my personal taste, but other than
that it is a well-maintained crate with a large ecosystem buy-in, and is
a solid bet.

The following pattern is suggested for using tracing:

```rust
  fn some_function(arg: Arg) {
    let span = tracing::info_span!("some_function");
    let _span = span.enter();

    let foo = {
        let span = tracing::info_span!("some_function/foo");
        let _span = span.enter();

        complex_computation_a() + long_computation_b()
    }
  }
```

That is:

* use `info` level (no strong motivation here, maybe we should switch to
  debug?)
* use function name as the message (helps finding it using "goto
  symbol" in an IDE)
* use `func_name/section` format to instrument sections of code, but
  prefer to extract them into a separate function.
* do not use `#[instrument]` attribute (it requires all args to be
  `Debug`, and works nicely only for common case).

### CODE REVIEW

File: runtime/near-vm-runner/src/cache.rs [hunk 1]
```diff
@@ -17,6 +17,9 @@ pub(crate) fn compile_module(
     code: &[u8],
     config: &VMConfig,
 ) -> Result<wasmer_runtime::Module, VMError> {
+    let span = tracing::info_span!("compile_module");
+    let _span = span.enter();
```
Comment (ailisp) [inline runtime/near-vm-runner/src/cache.rs:21]: Can this be simplified to `tracing::info_span!("compile_module").enter();`

**Review by ailisp — APPROVED** [review]
Beautiful API and clearly show which function is called by which one, nice!

**Review by bowenwang1996 — COMMENTED** [review]
@matklad should we extract `tracing_timings.rs` and put it in `tools` for example? We can also potentially replace delay detector with this. cc @pmnoxx

**Comment by willemneal** [comment]
I agree with Bowen, this seems like the start of a more general purpose tool which I could be used for code compiled to Wasm.

**Comment by pmnoxx** [comment]
> (builds on top of #3973)
> 
> `tracing` is a generic instrumentation framework. It allows manually
> instrumented code to signal point-in-time events and nested spans.
> Separately, a consumer for those spans and events can be defined.
> 
> In this particular case, we use `tracing` to measure time of various
> phases of contract execution, with the output ending up like this (read
> it bottom up):
> 
> ```
>     398.15ms compile_module
>   398.19ms compile_module_cached_wasmer_impl
>   140.19µs run_wasmer/call
> 407.21ms run_wasmer
> ```
> 
> The output's format can certainly be improved, but the beauty of
> `tracing` is that it can be done separately from instrumentation.
> 
> `tracing` is a bit over-engineered to my personal taste, but other than
> that it is a well-maintained crate with a large ecosystem buy-in, and is
> a solid bet.
> 
> The following pattern is suggested for using tracing:
> 
> ```rust
>   fn some_function(arg: Arg) {
>     let span = tracing::info_span!("some_function");
>     let _span = span.enter();
> 
>     let foo = {
>         let span = tracing::info_span!("some_function/foo");
>         let _span = span.enter();
> 
>         complex_computation_a() + long_computation_b()
>     }
>   }
> ```
> 
> That is:
> 
> * use `info` level (no strong motivation here, maybe we should switch to
>   debug?)
> * use function name as the message (helps finding it using "goto
>   symbol" in an IDE)
> * use `func_name/section` format to instrument sections of code, but
>   prefer to extract them into a separate function.
> * do not use `#[instrument]` attribute (it requires all args to be
>   `Debug`, and works nicely only for common case).

I was thinking about implementing something similar with a small different. I would like to integrate those stats with `utils/near-performance-metrics` and print them once a minute.

I would like to see something like this 
```

>     compile_module 398.15ms/30
>   compile_module_cached_wasmer_impl 398.19ms/15 
>   run_wasmer/call 140.19µs(199µs)/10 
> run_wasmer 407.21ms(350ms)/2
```
We could print both time data in following format avg time(max time)/number of calls.
I was going to implement something like for the purpose of investigating how much times is spend in different methods.

File: runtime/near-vm-runner/src/cache.rs (hunk 1 above)
Comment (matklad) [reply runtime/near-vm-runner/src/cache.rs:21]: hopefully soon: https://github.com/tokio-rs/tracing/pull/1252!

**Comment by matklad** [comment]
>should we extract tracing_timings.rs and put it in tools for example?

@bowenwang1996 I prefer to let the general purpose bits mature near the call-site and abstracting them away after they are sufficiently ready. In particular, creating a whole crate for a single file seems overkill. Long-term, I think this'll be moved directly to a separate repo & crates.io crate, or maybe even folded directly with https://github.com/davidbarsky/tracing-tree. 

>I was going to implement something like for the purpose of investigating how much times is spend in different methods.

@pmnoxx I would suggest taking a look at the https://tracing.rs/tracing/, I think their infra is right for  this task. In particular, with tracing you can easily vary the consumers of the events streams. This is useful, because I imagine we want different consumers for "monitoring in production" and "debugging performance problems locally".

**Review by frol — APPROVED** [review]

**Review by olonho — APPROVED** [review]

**Pull request merged**